File: tree/heap/heap.c
/*
 * 树算法 - 最大堆/最小堆实现 (MaxHeap/MinHeap)
 * 完全二叉树，满足堆序性质
 * 用于优先队列、堆排序等
 */

#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

#define MAX_SIZE 100

// 堆类型枚举
typedef enum {
    MIN_HEAP,
    MAX_HEAP
} HeapType;

// 堆结构
typedef struct {
    int data[MAX_SIZE];
    int size;
    HeapType type;
} Heap;

/**
 * 创建堆
 */
Heap* create_heap(HeapType type) {
    Heap* heap = (Heap*)malloc(sizeof(Heap));
    heap->size = 0;
    heap->type = type;
    return heap;
}

/**
 * 交换两个元素
 */
void swap(int* a, int* b) {
    int temp = *a;
    *a = *b;
    *b = temp;
}

/**
 * 获取父节点索引
 */
int parent(int i) {
    return (i - 1) / 2;
}

/**
 * 获取左子节点索引
 */
int left_child(int i) {
    return 2 * i + 1;
}

/**
 * 获取右子节点索引
 */
int right_child(int i) {
    return 2 * i + 2;
}

/**
 * 比较函数，根据堆类型返回比较结果
 */
int compare(Heap* heap, int a, int b) {
    if (heap->type == MIN_HEAP) {
        return a < b;  // 最小堆：小的优先
    } else {
        return a > b;  // 最大堆：大的优先
    }
}

/**
 * 向上调整堆（插入时使用）
 */
void heapify_up(Heap* heap, int index) {
    while (index > 0 && compare(heap, heap->data[index], heap->data[parent(index)])) {
        swap(&heap->data[index], &heap->data[parent(index)]);
        index = parent(index);
    }
}

/**
 * 向下调整堆（删除时使用）
 */
void heapify_down(Heap* heap, int index) {
    int smallest_or_largest = index;
    int left = left_child(index);
    int right = right_child(index);
    
    if (left < heap->size && compare(heap, heap->data[left], heap->data[smallest_or_largest])) {
        smallest_or_largest = left;
    }
    
    if (right < heap->size && compare(heap, heap->data[right], heap->data[smallest_or_largest])) {
        smallest_or_largest = right;
    }
    
    if (smallest_or_largest != index) {
        swap(&heap->data[index], &heap->data[smallest_or_largest]);
        heapify_down(heap, smallest_or_largest);
    }
}

/**
 * 插入元素
 */
int heap_insert(Heap* heap, int value) {
    if (heap->size >= MAX_SIZE) {
        printf("堆已满，无法插入\n");
        return 0;
    }
    
    heap->data[heap->size] = value;
    heapify_up(heap, heap->size);
    heap->size++;
    
    return 1;
}

/**
 * 获取堆顶元素（最小值或最大值）
 */
int heap_peek(Heap* heap) {
    if (heap->size == 0) {
        printf("堆为空\n");
        return (heap->type == MIN_HEAP) ? INT_MAX : INT_MIN;
    }
    
    return heap->data[0];
}

/**
 * 删除堆顶元素
 */
int heap_extract(Heap* heap) {
    if (heap->size == 0) {
        printf("堆为空，无法删除\n");
        return (heap->type == MIN_HEAP) ? INT_MAX : INT_MIN;
    }
    
    int root = heap->data[0];
    heap->data[0] = heap->data[heap->size - 1];
    heap->size--;
    heapify_down(heap, 0);
    
    return root;
}

/**
 * 检查堆是否为空
 */
int heap_is_empty(Heap* heap) {
    return heap->size == 0;
}

/**
 * 获取堆的大小
 */
int heap_size(Heap* heap) {
    return heap->size;
}

/**
 * 打印堆的内容
 */
void print_heap(Heap* heap) {
    printf("堆内容 (");
    printf(heap->type == MIN_HEAP ? "最小堆" : "最大堆");
    printf("): [");
    
    for (int i = 0; i < heap->size; i++) {
        printf("%d", heap->data[i]);
        if (i < heap->size - 1) printf(", ");
    }
    printf("]\n");
}

/**
 * 验证堆性质
 */
int verify_heap_property(Heap* heap, int index) {
    if (index >= heap->size) {
        return 1;  // 超出范围，满足性质
    }
    
    int left = left_child(index);
    int right = right_child(index);
    
    // 检查与左子节点的堆序性质
    if (left < heap->size) {
        if (heap->type == MIN_HEAP && heap->data[index] > heap->data[left]) {
            return 0;
        }
        if (heap->type == MAX_HEAP && heap->data[index] < heap->data[left]) {
            return 0;
        }
    }
    
    // 检查与右子节点的堆序性质
    if (right < heap->size) {
        if (heap->type == MIN_HEAP && heap->data[index] > heap->data[right]) {
            return 0;
        }
        if (heap->type == MAX_HEAP && heap->data[index] < heap->data[right]) {
            return 0;
        }
    }
    
    // 递归检查子树
    return verify_heap_property(heap, left) && verify_heap_property(heap, right);
}

/**
 * 堆排序演示
 */
void heap_sort_demo(int arr[], int n, int ascending) {
    printf("\n堆排序 (%s): ", ascending ? "升序" : "降序");
    
    // 创建临时堆
    Heap* temp_heap = create_heap(ascending ? MIN_HEAP : MAX_HEAP);
    
    // 插入所有元素
    for (int i = 0; i < n; i++) {
        heap_insert(temp_heap, arr[i]);
    }
    
    // 依次取出元素
    printf("[");
    for (int i = 0; i < n; i++) {
        int val = heap_extract(temp_heap);
        printf("%d", val);
        if (i < n - 1) printf(", ");
    }
    printf("]\n");
    
    free(temp_heap);
}

/**
 * 优先队列应用示例
 */
void priority_queue_demo() {
    printf("\n优先队列应用示例:\n");
    
    // 任务优先级队列（最大堆）
    Heap* task_queue = create_heap(MAX_HEAP);
    
    struct {
        int priority;
        char* description;
    } tasks[] = {
        {3, "低优先级任务"},
        {5, "高优先级任务"},
        {2, "很低优先级任务"},
        {4, "中优先级任务"},
        {1, "最低优先级任务"}
    };
    
    int num_tasks = sizeof(tasks) / sizeof(tasks[0]);
    
    // 添加任务
    printf("  添加任务:\n");
    for (int i = 0; i < num_tasks; i++) {
        heap_insert(task_queue, tasks[i].priority);
        printf("    添加: %s (优先级: %d)\n", 
               tasks[i].description, tasks[i].priority);
    }
    
    // 处理任务（按优先级）
    printf("  处理任务顺序:\n");
    int order = 1;
    while (!heap_is_empty(task_queue)) {
        int priority = heap_extract(task_queue);
        printf("    第%d个: 优先级 %d\n", order++, priority);
    }
    
    free(task_queue);
}

// 测试示例
int main() {
    printf("==================================================\n");
    printf("最大堆/最小堆实现 (MaxHeap/MinHeap)\n");
    printf("==================================================\n");
    
    // 最小堆测试
    printf("\n最小堆测试:\n");
    Heap* min_heap = create_heap(MIN_HEAP);
    
    int min_data[] = {5, 3, 8, 1, 6, 9, 2, 7, 4};
    int min_size = sizeof(min_data) / sizeof(min_data[0]);
    
    for (int i = 0; i < min_size; i++) {
        heap_insert(min_heap, min_data[i]);
        printf("  插入 %d: ", min_data[i]);
        print_heap(min_heap);
    }
    
    printf("\n  堆顶元素: %d\n", heap_peek(min_heap));
    printf("  堆性质验证: %s\n", 
           verify_heap_property(min_heap, 0) ? "✓" : "✗");
    
    printf("\n  依次删除元素:\n");
    while (!heap_is_empty(min_heap)) {
        int val = heap_extract(min_heap);
        printf("    删除 %d: ", val);
        print_heap(min_heap);
    }
    
    // 最大堆测试
    printf("\n最大堆测试:\n");
    Heap* max_heap = create_heap(MAX_HEAP);
    
    int max_data[] = {5, 3, 8, 1, 6, 9, 2, 7, 4};
    int max_size = sizeof(max_data) / sizeof(max_data[0]);
    
    for (int i = 0; i < max_size; i++) {
        heap_insert(max_heap, max_data[i]);
    }
    
    printf("  最终堆: ");
    print_heap(max_heap);
    printf("  堆性质验证: %s\n", 
           verify_heap_property(max_heap, 0) ? "✓" : "✗");
    
    // 堆排序演示
    int sort_data[] = {64, 34, 25, 12, 22, 11, 90, 88, 45, 50};
    int sort_size = sizeof(sort_data) / sizeof(sort_data[0]);
    
    heap_sort_demo(sort_data, sort_size, 1);  // 升序
    heap_sort_demo(sort_data, sort_size, 0);  // 降序
    
    // 优先队列应用
    priority_queue_demo();
    
    printf("\n应用场景:\n");
    printf("  • 优先队列\n");
    printf("  • 堆排序\n");
    printf("  • 图算法 (Dijkstra, Prim)\n");
    printf("  • 操作系统任务调度\n");
    printf("  • 事件驱动模拟\n");
    printf("  • 查找第k大/小元素\n");
    
    printf("\n复杂度分析:\n");
    printf("  • 插入: O(log n)\n");
    printf("  • 删除堆顶: O(log n)\n");
    printf("  • 查看堆顶: O(1)\n");
    printf("  • 建堆: O(n)\n");
    printf("  • 空间复杂度: O(n)\n");
    
    // 释放内存
    free(min_heap);
    free(max_heap);
    
    return 0;
}

File: tree/heap/Heap.java
/**
 * Copyright © https://github.com/microwind All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */

/**
 *  树算法 - 最大堆/最小堆实现 (MaxHeap/MinHeap)
 * 完全二叉树满足堆序性质
 * 用于优先队列、堆排序等
 */

// 堆类型枚举
enum HeapType {
    MIN_HEAP, // 最小堆：父节点 <= 子节点
    MAX_HEAP // 最大堆：父节点 >= 子节点
}

// 堆结构
class Heap {
    private int[] data; // 存储堆元素的数组
    private int size; // 堆中元素的数量
    private HeapType type; // 堆类型（最小堆或最大堆）
    private static final int MAX_SIZE = 100; // 堆的最大容量
    
    public Heap(HeapType type) {
        this.data = new int[MAX_SIZE]; // 初始化数组
        this.size = 0; // 初始大小为0
        this.type = type; // 设置堆类型
    }
    
    /**
     * 获取父节点索引
     * 根据完全二叉树的性质，父节点索引 = (i - 1) / 2
     */
    private int parent(int i) {
        return (i - 1) / 2;
    }
    
    /**
     * 获取左子节点索引
     * 根据完全二叉树的性质，左子节点索引 = 2 * i + 1
     */
    private int leftChild(int i) {
        return 2 * i + 1;
    }
    
    /**
     * 获取右子节点索引
     * 根据完全二叉树的性质，右子节点索引 = 2 * i + 2
     */
    private int rightChild(int i) {
        return 2 * i + 2;
    }
    
    /**
     * 交换两个元素
     */
    private void swap(int i, int j) {
        int temp = data[i];
        data[i] = data[j];
        data[j] = temp;
    }
    
    /**
     * 向上堆化操作
     */
    private void heapifyUp(int index) {
        while (index > 0) {
            int parent = parent(index);
            
            boolean needSwap = false;
            if (type == HeapType.MIN_HEAP) {
                needSwap = data[parent] > data[index];
            } else {
                needSwap = data[parent] < data[index];
            }
            
            if (needSwap) {
                swap(parent, index);
                index = parent;
            } else {
                break;
            }
        }
    }
    
    /**
     * 向下堆化操作
     */
    private void heapifyDown(int index) {
        while (true) {
            int left = leftChild(index);
            int right = rightChild(index);
            int target = index;
            
            if (left < size) {
                boolean needSwap = false;
                if (type == HeapType.MIN_HEAP) {
                    needSwap = data[left] < data[target];
                } else {
                    needSwap = data[left] > data[target];
                }
                if (needSwap) {
                    target = left;
                }
            }
            
            if (right < size) {
                boolean needSwap = false;
                if (type == HeapType.MIN_HEAP) {
                    needSwap = data[right] < data[target];
                } else {
                    needSwap = data[right] > data[target];
                }
                if (needSwap) {
                    target = right;
                }
            }
            
            if (target != index) {
                swap(index, target);
                index = target;
            } else {
                break;
            }
        }
    }
    
    /**
     * 插入元素
     */
    public void insert(int value) {
        if (size >= MAX_SIZE) {
            throw new RuntimeException("Heap is full");
        }
        
        data[size] = value;
        heapifyUp(size);
        size++;
    }
    
    /**
     * 提取根元素
     */
    public int extract() {
        if (size == 0) {
            throw new RuntimeException("Heap is empty");
        }
        
        int root = data[0];
        data[0] = data[size - 1];
        size--;
        heapifyDown(0);
        
        return root;
    }
    
    /**
     * 查看根元素
     */
    public int peek() {
        if (size == 0) {
            throw new RuntimeException("Heap is empty");
        }
        return data[0];
    }
    
    /**
     * 获取堆大小
     */
    public int getSize() {
        return size;
    }
    
    /**
     * 检查堆是否为空
     */
    public boolean isEmpty() {
        return size == 0;
    }
    
    /**
     * 堆排序
     */
    public void heapSort(int[] arr, boolean ascending) {
        // build heap
        Heap heap = new Heap(ascending ? HeapType.MAX_HEAP : HeapType.MIN_HEAP);
        for (int value : arr) {
            heap.insert(value);
        }
        
        // 提取元素
        for (int i = 0; i < arr.length; i++) {
            arr[i] = heap.extract();
        }
    }
    
    /**
     * 查找第k大元素
     */
    public static int findKthLargest(int[] arr, int k) {
        Heap minHeap = new Heap(HeapType.MIN_HEAP);
        
        for (int i = 0; i < k; i++) {
            minHeap.insert(arr[i]);
        }
        
        for (int i = k; i < arr.length; i++) {
            if (arr[i] > minHeap.peek()) {
                minHeap.extract();
                minHeap.insert(arr[i]);
            }
        }
        
        return minHeap.peek();
    }
    
    /**
     * 中位数查找器
     */
    public static void medianFinder() {
        System.out.println("Median Finder Demo:");
        int[] data = {5, 15, 1, 3, 8, 7, 9, 10, 6, 11, 4};
        Heap maxHeap = new Heap(HeapType.MAX_HEAP);
        Heap minHeap = new Heap(HeapType.MIN_HEAP);
        
        for (int num : data) {
            System.out.printf("Insert %d: ", num);
            
            if (maxHeap.isEmpty() || num <= maxHeap.peek()) {
                maxHeap.insert(num);
            } else {
                minHeap.insert(num);
            }
            
            // balance heaps
            if (maxHeap.getSize() > minHeap.getSize() + 1) {
                minHeap.insert(maxHeap.extract());
            } else if (minHeap.getSize() > maxHeap.getSize()) {
                maxHeap.insert(minHeap.extract());
            }
            
            // calculate median
            double median;
            if (maxHeap.getSize() == minHeap.getSize()) {
                median = (maxHeap.peek() + minHeap.peek()) / 2.0;
            } else {
                median = maxHeap.peek();
            }
            
            System.out.printf("Median = %.1f\n", median);
        }
    }
    
    /**
     * 主函数 - 测试堆
     */
    public static void main(String[] args) {
        System.out.println("=".repeat(50));
        System.out.println("Heap Algorithm - Java Implementation");
        System.out.println("=".repeat(50));
        
        // test max heap
        System.out.println("\n=== Max Heap Test ===");
        Heap maxHeap = new Heap(HeapType.MAX_HEAP);
        int[] testData = {4, 10, 3, 5, 1};
        
        System.out.println("Insert elements:");
        for (int value : testData) {
            maxHeap.insert(value);
            System.out.printf("Insert %d: ", value);
            printHeap(maxHeap);
        }
        
        System.out.println("\nExtract elements:");
        while (!maxHeap.isEmpty()) {
            int extracted = maxHeap.extract();
            System.out.printf("Extract %d: ", extracted);
            printHeap(maxHeap);
        }
        
        // test min heap
        System.out.println("\n=== Min Heap Test ===");
        Heap minHeap = new Heap(HeapType.MIN_HEAP);
        
        System.out.println("Insert elements:");
        for (int value : testData) {
            minHeap.insert(value);
            System.out.printf("Insert %d: ", value);
            printHeap(minHeap);
        }
        
        // 测试堆排序
        System.out.println("\n=== Heap Sort Test ===");
        int[] sortData = {3, 1, 4, 1, 5, 9, 2, 6};
        System.out.println("Original: " + java.util.Arrays.toString(sortData));
        
        int[] ascending = sortData.clone();
        maxHeap.heapSort(ascending, true);
        System.out.println("Ascending: " + java.util.Arrays.toString(ascending));
        
        int[] descending = sortData.clone();
        maxHeap.heapSort(descending, false);
        System.out.println("Descending: " + java.util.Arrays.toString(descending));
        
        // test kth largest
        System.out.println("\n=== Kth Largest Test ===");
        int[] kthData = {3, 2, 1, 5, 6, 4};
        System.out.println("Array: " + java.util.Arrays.toString(kthData));
        for (int k = 1; k <= kthData.length; k++) {
            int kth = findKthLargest(kthData, k);
            System.out.printf("%dth largest: %d\n", k, kth);
        }
        
        // 测试中位数查找器
        medianFinder();
        
        System.out.println("\n=== Heap Characteristics ===");
        System.out.println("1. 完全二叉树，存储在数组中");
        System.out.println("2. 最小堆：父节点 <= 子节点");
        System.out.println("3. 最大堆：父节点 >= 子节点");
        System.out.println("4. 插入和提取时间复杂度：O(log n)");
        System.out.println("5. 查看堆顶时间复杂度：O(1)");
        System.out.println("6. 堆排序时间复杂度：O(n log n)");
        System.out.println("7. 空间复杂度：O(n)");
        
        System.out.println("\n=== 应用 ===");
        System.out.println("优先队列、堆排序、查找第k大元素,");
        System.out.println("数据流中位数、图算法（Dijkstra、Prim）,");
        System.out.println("操作系统任务调度、事件模拟。");
    }
    
    /**
     * 打印堆元素
     */
    private static void printHeap(Heap heap) {
        System.out.print("[");
        for (int i = 0; i < heap.getSize(); i++) {
            System.out.print(heap.data[i]);
            if (i < heap.getSize() - 1) {
                System.out.print(", ");
            }
        }
        System.out.println("]");
    }
}
